synapse/handlers/federation.py

```diff
@@ -858,6 +858,49 @@ def maybe_backfill(self, room_id, current_depth):
             logger.debug("Not backfilling as no extremeties found.")
             return
 
+        # We only want to paginate if we can actually see the events we'll get,
+        # as otherwise we'll just spend a lot of resources to get redacted
+        # events.
+        #
+        # We do this by filtering all the extremities and seeing if any remain.
+        # Given we don't have the extremity events themselves, we need to
+        # actually check the events that reference them.
+        #
+        # *Note*: the spec wants us to keep backfilling until we reach the start
+        # of the room in case we are allowed to see some of the history. However
+        # in practice that causes more issues than its worth, as a) its
+        # relatively rare for there to be any visible history and b) even when
+        # there is its often sufficiently long ago that clients would stop
+        # attempting to paginate before backfill reached the visible history.
+        #
+        # TODO: If we do do a backfill the we should filter the extremities to
+        #   only include those that point to visible portions of history.
+        #
+        # TODO: Correctly handle the case where we are allowed to see the
```

Comment:
how would we do this? is it a thing we are ever likely to be able to do? if not, a TODO is inappropriate

Reply:
It should be possible, I've updated the comment with roughly what's needed.